File: androidApp/src/main/java/com/hien/mykmm/android/ui/RocketViewModel.kt
package com.hien.mykmm.android.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hien.mykmm.data.SpaceXRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class RocketUiState(
    val rocketItems: List<RocketItemUiState> = listOf(),
    val externalLink: String? = null
)

data class RocketItemUiState(
    val mission: String,
    val launchYear: String,
    val flightNumber: String,
    val details: String?,
    val article: String?
)

class RocketViewModel(
    private val spaceXRepository: SpaceXRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(RocketUiState())
    val uiState: StateFlow<RocketUiState> = _uiState.asStateFlow()

    init {
        viewModelScope.launch {
            val items: List<RocketItemUiState> = spaceXRepository.getAllLaunches().map {
                RocketItemUiState(
                    mission = it.missionName,
                    launchYear = it.launchYear.toString(),
                    flightNumber = it.flightNumber.toString(),
                    details = it.details,
                    article = it.links.article
                )
            }
            _uiState.update {
                it.copy(rocketItems = items)
            }
        }
    }

    fun onItemClicked(article: String?) {
        _uiState.update {
            it.copy(externalLink = article)
        }
    }

    fun resetExternalLink() {
        _uiState.update {
            it.copy(externalLink = null)
        }
    }
}

File: iosApp/iosApp/RocketScreen/RocketViewModel.swift
//
//  RocketViewModel.swift
//  iosApp
//
//  Created by Hien Le on 01.03.23.
//  Copyright © 2023 orgName. All rights reserved.
//

import shared

struct RocketUiState {
    var rocketItems: [RocketItemUiState] = []
}

struct RocketItemUiState {
    let mission: String
    let launchYear: String
    let flightNumber: String
    let details: String?
}

class RocketViewModel: ObservableObject {
    @Published var uiState: RocketUiState = RocketUiState()
    
    func loadItems() async {
        do {
            let items = try await SpaceXRepositoryHelper().getAllLaunches()
            let rocketItems: [RocketItemUiState] = items.map { item in
                return RocketItemUiState(mission: item.missionName, launchYear: String(item.launchYear), flightNumber: String(item.flightNumber), details: item.details)
            }
            DispatchQueue.main.async {
                self.uiState.rocketItems = rocketItems
            }
        } catch {
            print("Error loading launches: \(error)")
        }
    }
}
